Add Javalin functional tests

NicMcPhee opened the pull request:
This follows the example from ["Testing Javalin Applications"](https://javalin.io/tutorials/testing) for adding functional tests to a Javalin app. This does give us very nearly 100% test coverage, including coverage of the `Server` class, which was entirely uncovered before.

The tests in `ServerFunctionalSpec` are a little odd, mostly because the examples that they test are unusual (or, in the case of "/hello", totally unnecessary).

One could essentially replace all/most of the unit tests with functional tests like these, and still get quite reasonable coverage. That wouldn't be the recommended approach, but it would probably be where students would gravitate since it's often easier to think about "big picture" E2E-style tests.

Is this something we want to do? I'm not entirely sure, TBH. I kinda think it may be, but I only just hammered this out, so I'm not sure how I feel about it and would love some feedback.

NicMcPhee committed:
Add Javalin functional tests

This follows the example from "Testing Javalin Applications" (https://javalin.io/tutorials/testing) for adding functional tests to a Javalin app. This does give us very nearly 100% test coverage, including coverage of the `Server` class, which was entirely uncovered before.

The tests in `ServerFunctionalSpec` are a little odd, mostly because the examples that they test are unusual (or, in the case of "/hello", totally unnecessary).

One could essentially replace all/most of the unit tests with functional tests like these, and still get quite reasonable coverage. That wouldn't be the recommended approach, but it would probably be where students would gravitate since it's often easier to think about "big picture" E2E-style tests.

NicMcPhee commented on server/build.gradle:
```diff
@@ -86,7 +92,7 @@ jacocoTestReport {
     // leaving it out of the coverage report and the coverage limits.
     classDirectories.setFrom(files(classDirectories.files.collect {
-      fileTree(dir: it, exclude: 'umm3601/Server.class')
+      fileTree(dir: it)
```
We don't have to exclude the `Server` class anymore.

NicMcPhee commented on server/src/main/java/umm3601/Server.java:
```diff
@@ -12,13 +12,14 @@
   public static final String CLIENT_DIRECTORY = "../client";
   public static final String USER_DATA_FILE = "/users.json";
   private static UserDatabase userDatabase;
+  private static Javalin server;
```
I needed to pull `server` out to be a field so I could write `stopServer()` below. `stopServer()` was needed so that tests that started the server could stop it when they were done to free up the port, allowing other tests to start the server up again.

We probably want to pull a lot of what's in `main` out into something like `startServer()` method, and maybe even create a `configureServer` to put all the `get()`, etc., methods in.

NicMcPhee commented on server/src/main/java/umm3601/user/UserController.java:
```diff
@@ -33,7 +34,7 @@ public void getUser(Context ctx) {
     if (user != null) {
       ctx.json(user);
-      ctx.status(201);
+      ctx.status(HttpCode.OK);
```
It turns out that Javalin has an `HttpCode` class that has all the relevant response codes, and works nicely with Javalin in some useful ways. I probably want to use this instead of the `java.net` class that I used in the iteration template.

helloworld12321 replied:
I wouldn't worry about it—[those][1] [status][2] [code][3] [enums][4] [are][5] [a][6] [dime][7] [a][8] [dozen][9]

We're lucky that we're only using \*two\* different ones 😜

[1]: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/net/HttpURLConnection.html
[2]: https://docs.oracle.com/javaee/7/api/javax/ws/rs/core/Response.Status.html
[3]: https://docs.oracle.com/javaee/7/api/javax/servlet/http/HttpServletResponse.html
[4]: https://hc.apache.org/httpcomponents-core-5.1.x/current/httpcore5/apidocs/org/apache/hc/core5/http/HttpStatus.html
[5]: https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/HttpStatus.html
[6]: https://github.com/tipsy/javalin/blob/master/javalin/src/main/java/io/javalin/http/HttpCode.kt
[7]: https://htmlunit.sourceforge.io/apidocs/com/gargoylesoftware/htmlunit/WebResponse.html
[8]: https://www.eclipse.org/jetty/javadoc/jetty-11/org/eclipse/jetty/http/HttpStatus.Code.html
[9]: https://netty.io/4.0/api/io/netty/handler/codec/http/HttpResponseStatus.html

NicMcPhee committed:
Merge branch 'main' into add-functional-tests

NicMcPhee committed:
Merge branch 'main' into add-functional-tests

NicMcPhee committed:
Merge branch 'main' into add-functional-tests

NicMcPhee committed:
Merge branch 'main' into add-functional-tests

helloworld12321 commented on server/src/test/java/umm3601/user/UserFunctionalSpec.java:
```diff
@@ -0,0 +1,65 @@
+package umm3601.user;
```
I might suggest putting functional tests in the top-level `umm3601` package, since they test the whole server at once! (That also separates them conceptually from the unit tests, which may help the students.)

NicMcPhee replied:
I did put some at that level, but put these here because they were specific to users. I'll have to think about that.

NicMcPhee replied:
OK – I'm convinced (for now). I'm moving them up and we'll see how that feels. I almost wonder if what we want is a parallel directory structure `functional` alongside `main` and `test`?

helloworld12321 replied:
Oh, yeah, that sounds great. (I haven't seen that exact layout, but I've seen similar things in the past! 🙂)

NicMcPhee replied:
I'm not sure that I'll do that here, but I'll keep it in mind for later. We'll see how it goes having them in the `umm3601` package for now.

helloworld12321 replied:
I like that idea in theory! Unfortunately I don't think `User` has an appropriately-defined `.equals()` method.

What we could do instead is something like…

```java
String id = "588935f5c668650dc77df581";
JsonNode expectedUser = objectMapper.valueToTree(userDatabase.getUser(id));

HttpResponse<String> response = Unirest.get("http://localhost:4567/api/users/" + id).asString();
JsonNode receivedUser = objectMapper.readTree(response.getBody());

assertEquals(HttpCode.OK.getStatus(), response.getStatus(), "Getting an existing user should return OK status");
assertEquals(expectedUser, receivedUser, "The user didn't match");
```

helloworld12321 commented on server/src/test/java/umm3601/user/UserFunctionalSpec.java:
```diff
@@ -0,0 +1,65 @@
+    userDatabase = new UserDatabase(Server.USER_DATA_FILE);
+  }
+
+  @Test
+  public void canGetAllUsers() throws JsonProcessingException {
+    String usersJson = objectMapper.writeValueAsString(userDatabase.listUsers(new HashMap<String, List<String>>()));
+    HttpResponse<String> response = Unirest.get("http://localhost:4567/api/users").asString();
+    assertThat(response.getStatus()).isEqualTo(HttpCode.OK.getStatus());
+    assertThat(response.getBody()).isEqualTo(usersJson);
+  }
+
+  @Test
+  public void canGetSingleUserById() throws JsonProcessingException {
+    String id = "588935f5c668650dc77df581";
+    String user = objectMapper.writeValueAsString(userDatabase.getUser(id));
+    HttpResponse<String> response = Unirest.get("http://localhost:4567/api/users/" + id).asString();
+    assertThat(response.getStatus()).isEqualTo(HttpCode.OK.getStatus());
+    assertThat(response.getBody()).isEqualTo(user);
+  }
```
I like that these tests are a lot sparser than the unit tests. That way, we can rely on the unit tests to catch the nitty-gritty errors, while letting the functional tests handle big-picture issues (like whether we're doing HTTP right 🙂).

I am a bit worried that we're duplicating logic in tests and app code, like `objectMapper.writeValueAsString(userDatabase.getUser(…))`. I know it's a bit brittle, but would it make sense to write…?

```java
String user = "<json data straight from the file>";
```

helloworld12321 replied:
We could always make them static constants, like

```java
private static final String USER_ID = "588935f5c668650dc77df581";
private static final String USER_JSON = "<json data straight from the file>";
```

NicMcPhee replied:
I'm not sure I follow completely. I get that there's some potential duplication here, but I'm not sure quite what you mean by "json from the file" or how that improves anything.

helloworld12321 replied:
Oh, it's not a big deal, I was just thinking that we could cut out the call to `userDatabase` if we copied-and-pasted the JSON object we want directly from the database file. (Again, a bit brittle, but it gives us slightly more end-to-end-ness 🤷🏻‍♂️)

NicMcPhee replied:
Ah, I get it. Given that the IDs _only_ make sense in the context of the "database" I'm not sure it's too weird to include the DB in the tests. What this probably suggests is that we (mostly) want tests that are less dependent on the contents of the DB, but that's not happening today.

helloworld12321 commented on server/src/test/java/umm3601/user/UserFunctionalSpec.java:
```diff
@@ -0,0 +1,65 @@
+    String usersJson = objectMapper.writeValueAsString(userDatabase.listUsers(new HashMap<String, List<String>>()));
+    HttpResponse<String> response = Unirest.get("http://localhost:4567/api/users").asString();
+    assertThat(response.getStatus()).isEqualTo(HttpCode.OK.getStatus());
+    assertThat(response.getBody()).isEqualTo(usersJson);
```
We likely want instances of `JsonNode` instead of comparing strings here.

Right now the test could fail if `ctx.json(…)` and `objectMapper.writeValueAsString(…)` use different serialization algorithms. (For example, one might alphabetize object properties, and the other doesn't; or one could put spaces after commas, and the other doesn't.)

I found this tutorial, in case it helps! 🙂 <https://www.baeldung.com/jackson-compare-two-json-objects>

NicMcPhee replied:
This was an important catch and a big help – thanks!

NicMcPhee committed:
Merge branch 'main' into add-functional-tests

helloworld12321 reviewed (changes requested):
This is neat stuff! 🙂 I've made a few suggestions, but the most important ones are:

1. Using AssertJ instead of vanilla JUnit might just confuse people. 🤷🏻‍♂️
2. In `UserFunctionalSpec`, we probably _don't_ want to compare JSON blobs with `String::equals`. (Jackson's `JsonNode::equals` method seems like it does what we want here.)

This is assuming we want functional tests at all—if we wanted to, we could probably get by with unit and e2e. But if functional tests are something we want the students to write, this looks like a good way to go about it! 🙂

helloworld12321 commented on server/src/test/java/umm3601/ServerFunctionalSpec.java:
```diff
@@ -0,0 +1,54 @@
+    assertThat(response.getBody()).isEqualTo("Hello World");
+  }
+
+  @Test
+  public void usersRedirects() {
+    HttpResponse<String> response = Unirest.get("http://localhost:4567/users").asString();
+    assertThat(response.getStatus()).isEqualTo(HttpCode.OK.getStatus());
+    assertThat(response.getBody()).contains("<!DOCTYPE html>");
+  }
+
+  @Test void todosRedirects() {
+    HttpResponse<String> response = Unirest.get("http://localhost:4567/todos").asString();
+    assertThat(response.getStatus()).isEqualTo(HttpCode.OK.getStatus());
+    assertThat(response.getBody()).contains("<!DOCTYPE html>");
+  }
```
These tests are fine as is, but what would you think about something like…? 

```java
import kong.unirest.HeaderNames;

// ...

public void staticFilesWork() {
  String[] staticFileEndpoints = {"users.html", "users", "todos.html", "todos"};
  for (String endpoint : staticFileEndpoints) {
    HttpResponse<String> response = Unirest.get("http://localhost:4567/" + endpoint).asString();
    assertThat(response.getStatus()).isEqualTo(HttpCode.OK.getStatus());
    assertThat(response.getHeaders().getFirst(HeaderNames.CONTENT_TYPE)).isEqualTo("text/html");
    assertThat(response.getBody()).contains("<!DOCTYPE html>");
  }
}
```

That way, we save some code duplication, check a few more endpoints, and make sure that the static files have the right MIME type 🙂

NicMcPhee commented:
All good points/questions.

I just used AssertJ because that's what they used in the example, which isn't much of a reason. 😳 I'll have to think about that.

Well spotted on the equality tests!

I do feel like these are potentially helpful, and in a different way than E2E tests. E2E tests require having and understanding both ends of the system, and you have to be able to manipulate the client to produce every possible REST request.

Here the tests are self-contained in the server side, making it easier to focus on that part of the problem, which could be particularly useful for this lab.

All that said, we start this lab tomorrow, and I won't have time to update the write-up, so I doubt it will happen now. Maybe lab 4?

NicMcPhee committed:
Update server/src/test/java/umm3601/ServerFunctionalSpec.java

Co-authored-by: Joe Moonan Walbran <walbr037@morris.umn.edu>

helloworld12321 commented:
@NicMcPhee Makes sense, thanks so much! Yes, that sounds great to me.

NicMcPhee committed:
Merge branch 'main' into add-functional-tests

NicMcPhee committed:
Merge branch 'main' into add-functional-tests

NicMcPhee committed:
Move user controller functional test file to umm3601 package

This moves `UserFuncitonalSpec.java` to the `umm3601` package, and updates the imports accordingly.

NicMcPhee committed:
Merge branch 'main' into add-functional-tests

NicMcPhee committed:
Remove the AssertJ dependency

This removes that dependency and the handful of uses of `assertThat`, replacing them with `assertEquals` or `assertTrue`.

It's possible that the use of `contains()` in `ServerFunctionalSpec` could be replaced by something fancy from AssertJ, but no need to worry about it here.

NicMcPhee committed:
Check all static files

This uses a nice refactoring proposed by @helloworld12321 (https://github.com/UMM-CSci-3601/3601-lab2_client-server/pull/224#discussion_r796276955) that now checks more things (all four static HTML links) in less code.

NicMcPhee committed:
Add comment about why we're comparing JSON instead of Strings

I added a comment to the two places where we use the Jackson ObjectMapper to compare JSON objects instead of the underlying Strings.

NicMcPhee committed:
Merge branch 'add-functional-tests' of https://github.com/UMM-CSci-3601/3601-lab2_client-server into add-functional-tests

helloworld12321 commented on server/src/test/java/umm3601/UserFunctionalSpec.java:
```diff
@@ -16,7 +16,7 @@
 import kong.unirest.HttpResponse;
 import kong.unirest.Unirest;
-import umm3601.Server;
+import umm3601.user.UserDatabase;
```
We need both these imports, I think.

```suggestion
import umm3601.Server;
import umm3601.user.UserDatabase;
```

NicMcPhee replied:
We don't need the `Server` import because we're now in the `umm3601` where `Server` lives. Everything in you package is automatically imported, and you only need to import things in other packages.

helloworld12321 replied:
Oh, maybe we don't if we're in the same package. Never mind, ignore me! 😅

helloworld12321 commented on server/src/test/java/umm3601/UserFunctionalSpec.java:
```diff
@@ -50,16 +50,32 @@ public static void createDbAndMapper() throws IOException {
-    assertThat(response.getStatus()).isEqualTo(HttpCode.OK.getStatus());
-    assertThat(response.getBody()).isEqualTo(user);
+    assertEquals(HttpCode.OK.getStatus(), response.getStatus(), "Getting an existing user should return OK status");
+    /*
+     * We don't want to just compare strings directly, because the order of fields in a JSON object isn't specified,
+     * and two "equal" objects could have their fields in different orders. If we just did a pure String comparison
+     * we'd get that wrong; if we use the Jackson ObjectMapper to turn them into JSON objects and compare *those*,
+     * we we avoid that problem.
+     *
+     * See this for more examples and info: https://www.baeldung.com/jackson-compare-two-json-objects
+     */
```
We probably only need this comment once (DRY and all that 🙂)
```suggestion
```

NicMcPhee replied:
Actually, I'm inclined to duplicate it (at least for now). Students often don't "read" code files top to bottom, and I've found that a comment that's somewhere else will often be missed.

helloworld12321 replied:
Oh, cool, got it 🙂

helloworld12321 reviewed (changes requested):
Looks pretty good! :) I'd be happy to ship it after just a few changes.

[Although see the discussion on slack about whether it makes sense to include functional tests.]

floogulinc commented on server/src/test/java/umm3601/ServerFunctionalSpec.java:
```diff
@@ -0,0 +1,53 @@
+
+  @Test
+  public void staticFilesWork() {
+    String[] staticFileEndpoints = { "users.html", "users", "todos.html", "todos" };
```
`/users` and `/todos` aren't HTML. They are redirects so this will fail.

helloworld12321 replied:
Hmm, I tested locally, and I _think_ Unirest follows redirects by default, but I need to check that.

NicMcPhee replied:
These tests do pass. I was somewhat surprised by that, but I don't think that Javalin is actually sending a redirect back to the browser. My poking around last night made it look like Javalin actually returns a 200 and sends the HTML directly. Is that what you were hoping it would do back when you added these @floogulinc?

floogulinc replied:
It doesn't send the HTML, it sends an HTTP 302 redirect. That's why when you visit one of those paths in the browser it becomes `users.html` for example. 

![image](https://user-images.githubusercontent.com/1300395/152038719-59e11de5-f20d-4d0a-9b60-e2292aceba7b.png)

If the tests are not failing here it could be like @helloworld12321 says and Unirest is following the redirect automatically which isn't actually great for testing.

helloworld12321 replied:
I'm sure it's configurable if we need it to be—In the past I've used [`HtmlUnit`](https://htmlunit.sourceforge.io) for this kind of testing and it had the same default behavior.

floogulinc commented on server/src/test/java/umm3601/ServerFunctionalSpec.java:
```diff
@@ -0,0 +1,53 @@
+
+  @Test
+  public void staticFilesWork() {
+    String[] staticFileEndpoints = { "users.html", "users", "todos.html", "todos" };
```
Shouldn't this also include `''` for the index page? `index.html` would also suffice.

NicMcPhee replied:
No reason not to. Should we have both `""` and `"index.html"`?

helloworld12321 commented on server/src/test/java/umm3601/ServerFunctionalSpec.java:
```diff
@@ -0,0 +1,53 @@
+    String[] staticFileEndpoints = { "users.html", "users", "todos.html", "todos" };
+    for (String endpoint : staticFileEndpoints) {
+      HttpResponse<String> response = Unirest.get("http://localhost:4567/" + endpoint).asString();
+      assertEquals(HttpCode.OK.getStatus(), response.getStatus(), "The '/users' path should return OK status");
```
Oh, dang, you're right. Sorry about that!

NicMcPhee replied:
Good catch – I'll change that.

helloworld12321 replied:
```suggestion
      assertEquals(
          HttpCode.OK.getStatus(),
          response.getStatus(),
          String.sprintf("The '%s' path should return OK status", endpoint));
```

NicMcPhee committed:
Address Checkstyle issues

I forgot to run Checkstyle on the last commit, and I'd introduced a few problems that caused the build to fail on GitHub.

Most of this was just long lines, but there is a fight going on between VS Code and Checkstyle about whether there should be spaces after `{` in initializing arrays. I actually like the space, so perhaps we want to disable or tweak the Checkstyle check.

helloworld12321 commented on server/src/test/java/umm3601/ServerFunctionalSpec.java:
```diff
@@ -0,0 +1,55 @@
+
+  @Test
+  public void staticFilesWork() {
+    String[] staticFileEndpoints = {"users.html", "users", "todos.html", "todos"};
```
```suggestion
    String[] staticFileEndpoints = {"users.html", "users", "todos.html", "todos", "index.html", ""};
```

NicMcPhee replied:
I went ahead and added both `""` and `"index.html"`, and put them at the front of the list instead of the end.

floogulinc commented on server/src/test/java/umm3601/UserFunctionalSpec.java:
```diff
@@ -0,0 +1,81 @@
+     * See this for more examples and info: https://www.baeldung.com/jackson-compare-two-json-objects
+     */
+    assertEquals(objectMapper.readTree(user), objectMapper.readTree(response.getBody()), "The user didn't match");
+  }
```
Since you are converting the `user` back to an object here, wouldn't it make sense to just keep it as an object the whole time. Eg not doing `objectMapper.writeValueAsString` in the first place?

Also Unirest can directly turn JSON into an object (see https://kong.github.io/unirest-java/#responses) so you could simplify this whole thing to something like:

```java
String id = "588935f5c668650dc77df581";
HttpResponse<String> response = Unirest.get("http://localhost:4567/api/users/" + id).asObject(User.class);

assertEquals(HttpCode.OK.getStatus(), response.getStatus(), "Getting an existing user should return OK status");

assertEquals(userDatabase.getUser(id), response.getBody(), "The user didn't match");
```

Similar for `canGetAllUsers` above.

NicMcPhee replied:
Good catch. That had been niggling at me while I was doing that last night, but I think I was too deep in the woods to see the trees.

NicMcPhee replied:
Except, that requires having a meaningful `User#equals()` method, which we don't have. Ugh.

NicMcPhee replied:
Do we think it's reasonable to just have `User#equals()` only check that the IDs match, or should we check all the fields?

helloworld12321 replied:
We could always do something like…

```java
String id = "588935f5c668650dc77df581";
JsonNode expectedUser = objectMapper.readValueAsTree(userDatabase.getUser(id));

HttpResponse<String> response = Unirest.get("http://localhost:4567/api/users/" + id).asString();
JsonNode receivedUser = objectMapper.readTree(response.getBody());

assertEquals(HttpCode.OK.getStatus(), response.getStatus(), "Getting an existing user should return OK status");
/*
 * We don't want to just compare strings directly, because the order of fields in a JSON object isn't specified,
 * and two "equal" objects could have their fields in different orders. If we just did a pure String comparison
 * we'd get that wrong; if we use the Jackson ObjectMapper to turn them into JSON objects and compare *those*,
 * we we avoid that problem.
 *
 * See this for more examples and info: https://www.baeldung.com/jackson-compare-two-json-objects
 */
assertEquals(expectedUser, receivedUser, "The user didn't match");
```

floogulinc replied:
If you were to do that I would suggest instead just doing that only in the test, eg:

```java
assertEquals(userDatabase.getUser(id)._id, response.getBody()._id, "The user didn't match");
```

NicMcPhee committed:
Add "" and "index.html" to the list of static paths

This adds the two other static paths to the `ServerFunctionalSpec#staticFilesWork()` test.

floogulinc commented:
As a general thing, this will mean the tests can't be run while you are also running the server normally like you can right now. I suggest adding some detection or command flag to not start the javalin server when running the tests in the case you are already running it separately.

NicMcPhee commented:
That's a great point, @floogulinc, and I totally hadn't thought of that. I think it may be useful to step away from this PR for now, as there just seem to be too many questions that we don't really have time to address at the moment.

I don't know how easy/hard it would be to detect another server running. And can we tell if it's _our_ server, and not some other person's server that they left running on this box when they went to lunch earlier? If we can't, that's bad, since we sure don't want to think we're testing our server when we're actually testing theirs (and none of our changes will ever affect anything).

floogulinc commented:
@NicMcPhee an alternative I thought of could be to have the server in the tests run on a different port and since there is no actual DB in this lab it would probably be fine though that wouldn't necessarily work well in future labs with a real DB.

helloworld12321 commented:
I also wouldn't mind failing-fast if there's another server running ¯\\\_(ツ)\_/¯ That's probably the simplest optoin.

floogulinc commented:
Yeah that's fair, its just a bit disappointing to break the ability to be running all the dev servers at the same time as running your tests as students can currently do even in the full iteration with java and angular unit tests. Especially since the eventual e2e tests _require_ the server to be running separately. It seems odd that they would have to shut it down again just to run these Java tests.

NicMcPhee commented:
I like the idea of the functional tests just running on a different port. It requires some changes to how the server gets spun up, but that should probably be cleaned up anyway. And that way the server can be started (on the new port), the tests can be run, and then the server shut down, freeing that port up.

floogulinc commented:
The only issue I see with that is as soon as the actual Mongo DB is added you will have two instances of the server potentially talking to the DB at the same time.

NicMcPhee committed:
Remove JSON translation from functional tests

@floogulinc pointed out that we didn't really need to have all the JSON translation in the functional tests, and we could just compare the POJOs directly. This changes the tests to work on POJOs.

To make this work I had to implement an `equals()` method for the `User` class; otherwise testing for equality failed because the object we got from the DB was different than the object we got from the response. To keep things simple I implemented this to just test to see if the `_id`s were the same, but we could make that fancier and check the other fields as well.

Implementing `equals()` meant that I had to implement `hashCode()`, which just returns the hash of the ID string.

I also added a simple `toString()` that prints the name to aid in some debugging.

NicMcPhee commented:
Just for the record, we did _not_ bring this in for Lab 2 of the S22 iteration the class. We could still bring this forward into Lab 3, etc., but I'm not going to solve that problem now. @floogulinc has raised some important questions that would need addressing before we go there.

I'm actually going to convert this back to a draft PR to make it clear that it's not really ready for merging. I would still like to continue the discussion though (either here or on Slack) about whether we want to bring this into future labs, or future iterations of the course.

NicMcPhee commented:
After a lot of work and good discussion last year, this basically got dropped. @kklamberty starts this semester's edition of the course tomorrow, so we might want to come back to this while we still have some time to make some decisions, changes, etc., etc. There were definitely a host of unanswered/unresolved questions which should be attended to if we're going to try to move forward on this for S23.

Thoughts?

kklamberty commented:
There is an issue on the Iteration Template about using Unirest to do similar testing. https://github.com/UMM-CSci-3601/3601-iteration-template/issues/675

NicMcPhee commented:
Thanks to @kklamberty for noticing the similarities here. I haven't actually tried Unirest, and I don't know if it might simplify/address some of the concerns outlined above.

I might try a version with Unirest to see if it helps the port problem. Even if it does, though, we have the potential issue that @floogulinc raised of this wiping the test database. I'm a little less worried about that, since presumably the test database is there to be wiped out and rebuilt. So as long as we're not running these tests at the same time as some other tests, we should be good.

I also added [an issue to the iteration template](https://github.com/UMM-CSci-3601/3601-iteration-template/issues/1391) about whether we could/should using mocking/spies to actually eliminate the calls to Mongo in the server _unit_ tests. I'm not sure that we should, but I wanted to raise the issue.

NicMcPhee committed:
Merge branch 'main' into add-functional-tests

kklamberty commented:
I was just browsing through the issues and plans for software design, and came across this PR. I made some rather large changes to the lab structure this semester, and I'm not 100% sure if some of this/all of this would only be useful for the old lab 2 (there are some comments that suggest things would be different somehow with the Mongo stuff actually involved).
